### PR TITLE
rewrite Load plan

### PR DESCRIPTION
**Pull request by jianwan0214**
## What type of PR is this?

- [ ] API-change
- [ ] BUG
- [ ] Improvement
- [ ] Documentation
- [x] Feature
- [ ] Test and CI
- [ ] Code Refactoring

## Which issue(s) this PR fixes:

issue #3421 

## What this PR does / why we need it:
rewrite the load plan as insert into table select from external table;
add sql to support external table as we can select info from external table;

### CODE REVIEW

File: pkg/sql/plan/query_builder.go [hunk 1]
```diff
@@ -91,8 +92,10 @@ func (builder *QueryBuilder) remapAllColRefs(nodeID int32, colRefCnt map[[2]int3
 
 		tag := node.BindingTags[0]
 		newTableDef := &plan.TableDef{
-			Name: node.TableDef.Name,
-			Defs: node.TableDef.Defs,
+			Name:          node.TableDef.Name,
+			Defs:          node.TableDef.Defs,
+			Name2ColIndex: node.TableDef.Name2ColIndex,
```
Comment (ouyuanning) [inline pkg/sql/plan/query_builder.go:97]: how about add  an "Idx" field to ColDef ?  just like schema.ColDef

File: pkg/vm/pipeline/pipeline.go [hunk 2]
```diff
@@ -79,6 +87,47 @@ func (p *Pipeline) Run(r engine.Reader, proc *process.Process) (bool, error) {
...
+	if err != nil {
+		return false, err
+	}
+
+	fileList, err := getLoadDataList(load)
+	if err != nil {
+		return false, err
+	}
+
+	fmt.Println("wangjian sql1 is", load.IgnoredLines)
```
Comment (ouyuanning) [inline pkg/vm/pipeline/pipeline.go:116]: remove test code?

File: pkg/vm/pipeline/pipeline.go [hunk 3]
```diff
@@ -79,6 +87,47 @@ func (p *Pipeline) Run(r engine.Reader, proc *process.Process) (bool, error) {
 	}
 }
 
+func (p *Pipeline) Run2(proc *process.Process) (bool, error) {
```
Comment (ouyuanning) [inline pkg/vm/pipeline/pipeline.go:90]: maybe use another function name?

File: pkg/vm/pipeline/types.go [hunk 4]
```diff
@@ -79,8 +83,22 @@ import (
 //  -> π(c.city, s.city, revenue)
 type Pipeline struct {
 	// attrs, column list.
-	attrs []string
+	attrs         []string
+	cols          []*plan.ColDef
+	Name2ColIndex map[string]int32
```
Comment (ouyuanning) [inline pkg/vm/pipeline/types.go:88]: is private ok?

File: proto/plan.proto [hunk 5]
```diff
@@ -297,6 +297,8 @@ message TableDef {
 	string name				= 1;
 	repeated ColDef cols	= 2;
 	repeated DefType defs	= 3;
+	string TableType 		= 4;
```
Comment (ouyuanning) [inline proto/plan.proto:300]: use table_type in proto file.   
protobuf creator will deal with others

File: pkg/vm/pipeline/pipeline.go (hunk 3 above)
Comment (daviszhen) [inline pkg/vm/pipeline/pipeline.go:90]: Why do you need the `Run2`? Can you merge it with `Run`?

File: pkg/vm/pipeline/types.go [hunk 6]
```diff
@@ -79,8 +83,22 @@ import (
 //  -> π(c.city, s.city, revenue)
 type Pipeline struct {
 	// attrs, column list.
-	attrs []string
+	attrs         []string
```
Comment (daviszhen) [inline pkg/vm/pipeline/types.go:86]: Put it in an independent struct.
`
        attrs         []string
	cols          []*plan.ColDef
	Name2ColIndex map[string]int32
	CreateSql     string
`

File: pkg/sql/compile/scope.go [hunk 7]
```diff
@@ -64,6 +64,15 @@ func (s *Scope) Run(c *Compile) (err error) {
 	return nil
 }
 
+// Run read data from storage engine and run the instructions of scope.
+func (s *Scope) Run2(c *Compile) (err error) {
```
Comment (daviszhen) [inline pkg/sql/compile/scope.go:68]: Can you merge it with the 'Run'?

File: pkg/sql/compile/compile.go [hunk 8]
```diff
@@ -344,6 +344,9 @@ func (c *Compile) compilePlanScope(n *plan.Node, ns []*plan.Node) ([]*Scope, err
 	case plan.Node_TABLE_SCAN:
 		ss := c.compileTableScan(n)
 		return c.compileSort(n, c.compileProjection(n, c.compileRestrict(n, ss))), nil
+	case plan.Node_EXTERNAL_SCAN:
+		ss := c.compileExternalScan(n)
+		return c.compileSort(n, c.compileProjection(n, c.compileRestrict(n, ss))), nil
```
Comment (fengttt) [inline pkg/sql/compile/compile.go:349]: Another surprise, what is compileSort?   Why do we need sort?

File: pkg/sql/compile/scope.go [hunk 9]
```diff
@@ -101,6 +110,22 @@ func (s *Scope) MergeRun(c *Compile) error {
 				}()
 				err = cs.ParallelRun(c)
 			}(s.PreScopes[i])
+		case External:
```
Comment (fengttt) [inline pkg/sql/compile/scope.go:113]: All these *2 stuff, out of control.   Why do we need that, can we fix the *1 instead?

File: pkg/sql/compile/compile.go (hunk 8 above)
Comment (nnsgmsone) [reply pkg/sql/compile/compile.go:349]: External scan There will be no filter, projection, sorting or limit information at all, so why write it that way?

File: pkg/sql/compile/scope.go (hunk 9 above)
Comment (nnsgmsone) [reply pkg/sql/compile/scope.go:113]: Why is External a type, and isn't External supposed to be a specialization of ValueScan?

File: pkg/vm/pipeline/pipeline.go [hunk 10]
```diff
@@ -79,6 +87,47 @@ func (p *Pipeline) Run(r engine.Reader, proc *process.Process) (bool, error) {
...
+	var err error
```
Comment (nnsgmsone) [inline pkg/vm/pipeline/pipeline.go:91]: Please make the load a similar strategy to valuescan

**Review by nnsgmsone — CHANGES_REQUESTED** [review]

File: pkg/vm/pipeline/pipeline.go (hunk 10 above)
Comment (jianwan0214) [reply pkg/vm/pipeline/pipeline.go:91]: ok

File: proto/plan.proto (hunk 5 above)
Comment (ouyuanning) [inline proto/plan.proto:300]: table_type  and create_sql are items of defs.
may be you don't need add these field in TableDef。
when you use, you can get them in TableDef.Defs

File: pkg/vm/engine/tae/moengine/helper.go [hunk 11]
```diff
@@ -64,6 +64,19 @@ func SchemaToDefs(schema *catalog.Schema) (defs []engine.TableDef, err error) {
 		}
 		defs = append(defs, pk)
 	}
+	pro := new(engine.PropertiesDef)
+	pro.Properties = append(pro.Properties, engine.Property{
+		Key:   "relkind",
```
Comment (ouyuanning) [inline pkg/vm/engine/tae/moengine/helper.go:69]: relkind  -> catalog.SystemRelAttr_Kind

File: pkg/vm/engine/tae/moengine/helper.go [hunk 12]
```diff
@@ -64,6 +64,19 @@ func SchemaToDefs(schema *catalog.Schema) (defs []engine.TableDef, err error) {
...
+		Value: string(schema.Relkind),
+	})
+	if schema.Createsql != "" {
+		pro.Properties = append(pro.Properties, engine.Property{
+			Key:   "createsql",
```
Comment (ouyuanning) [inline pkg/vm/engine/tae/moengine/helper.go:74]: createsql -> SystemRelAttr_CreateSQL

File: pkg/vm/engine/tae/moengine/helper.go [hunk 13]
```diff
@@ -94,6 +107,10 @@ func DefsToSchema(name string, defs []engine.TableDef) (schema *catalog.Schema,
 			for _, property := range defVal.Properties {
 				if strings.ToLower(property.Key) == "comment" {
 					schema.Comment = property.Value
+				} else if strings.ToLower(property.Key) == "relkind" {
```
Comment (ouyuanning) [inline pkg/vm/engine/tae/moengine/helper.go:110]: can use switch  

switch property.key {
    case catalog.SystemRelAttr_Comment:
     ...
    case catalog.SystemRelAttr_Kind:
....
}

File: pkg/sql/plan/build_insert.go [hunk 14]
```diff
@@ -205,6 +207,7 @@ func MakeInsertError(id types.T, col *ColDef, rows []tree.Exprs, colIdx, rowIdx
 
 func buildInsertSelect(stmt *tree.Insert, ctx CompilerContext) (p *Plan, err error) {
 	pn, err := runBuildSelectByBinder(plan.Query_SELECT, ctx, stmt.Rows)
+	fmt.Println("wangjian sql7 is", pn)
```
Comment (iamlinjunhong) [inline pkg/sql/plan/build_insert.go:210]: remove

File: pkg/sql/plan/build_insert.go [hunk 15]
```diff
@@ -214,6 +217,7 @@ func buildInsertSelect(stmt *tree.Insert, ctx CompilerContext) (p *Plan, err err
 		return nil, errors.New(errno.InvalidColumnReference, "Column count doesn't match value count")
 	}
 
+	fmt.Println("wangjian sql8 is", stmt.Table)
```
Comment (iamlinjunhong) [inline pkg/sql/plan/build_insert.go:220]: remove

File: pkg/sql/plan/build_insert.go [hunk 16]
```diff
@@ -307,10 +311,12 @@ func getInsertExprs(stmt *tree.Insert, cols []*ColDef, tableDef *TableDef) ([]*E
 }
 
 func getInsertTable(stmt tree.TableExpr, ctx CompilerContext) (*ObjectRef, *TableDef, error) {
+	fmt.Printf("wangjian sql6b is %T\n", stmt)
```
Comment (iamlinjunhong) [inline pkg/sql/plan/build_insert.go:314]: remove

File: pkg/sql/plan/build_insert.go [hunk 17]
```diff
@@ -307,10 +311,12 @@ func getInsertExprs(stmt *tree.Insert, cols []*ColDef, tableDef *TableDef) ([]*E
...
 	switch tbl := stmt.(type) {
 	case *tree.TableName:
 		tblName := string(tbl.ObjectName)
 		dbName := string(tbl.SchemaName)
+		fmt.Println("wangjian sql6 is", dbName, tblName)
```
Comment (iamlinjunhong) [inline pkg/sql/plan/build_insert.go:319]: remove

**Review by iamlinjunhong — CHANGES_REQUESTED** [review]

**Review by daviszhen — APPROVED** [review]

File: pkg/sql/compile/scope.go [hunk 18]
```diff
@@ -51,10 +51,16 @@ func PrintScope(prefix []byte, ss []*Scope) {
 
 // Run read data from storage engine and run the instructions of scope.
 func (s *Scope) Run(c *Compile) (err error) {
-	p := pipeline.New(s.DataSource.Attributes, s.Instructions, s.Reg)
+	p := pipeline.New(s.DataSource.Attributes, s.Instructions, s.Reg, s.DataSource.ExternalParam)
```
Comment (nnsgmsone) [inline pkg/sql/compile/scope.go:54]: Why this parameter is needed

**Review by nnsgmsone — CHANGES_REQUESTED** [review]

File: pkg/sql/compile/scope.go (hunk 18 above)
Comment (jianwan0214) [reply pkg/sql/compile/scope.go:54]: I need this param to get the param like filepath

File: pkg/sql/plan/build_load.go [hunk 19]
```diff
@@ -0,0 +1,175 @@
...
+
+func buildLoad(stmt *tree.Load, ctx CompilerContext) (*Plan, error) {
+	tblName := string(stmt.Table.ObjectName)
+	dbName := string(stmt.Table.SchemaName)
+	objRef, tableDef := ctx.Resolve(dbName, tblName)
+	if tableDef == nil {
+		return nil, errors.New("", fmt.Sprintf("Invalid table name: %s", tree.String(stmt.Table, dialect.MYSQL)))
+	}
+
+	tableDef.Name2ColIndex = map[string]int32{}
```
Comment (ouyuanning) [inline pkg/sql/plan/build_load.go:35]: use appendNode method

Comment (ouyuanning) [reply pkg/sql/plan/build_load.go:35]: my mistake, there is not builder here.

File: pkg/sql/plan/build_load.go [hunk 20]
```diff
@@ -0,0 +1,175 @@
...
+			node.ProjectList[Name2ColIndex[v]] = projectVec[i]
+		}
+	}
+	for i := 0; i < len(tableDef.Cols); i++ {
+		if node.ProjectList[i] != nil {
+			continue
+		}
+		var tmp *plan.Expr
+		if tableDef.Cols[i].Default == nil || tableDef.Cols[i].Default.NullAbility {
+			tmp = &plan.Expr{
```
Comment (ouyuanning) [inline pkg/sql/plan/build_load.go:156]: use makePlan2NullConstExprWithType

File: pkg/sql/parsers/dialect/mysql/mysql_sql.y [hunk 21]
```diff
@@ -3876,6 +3861,92 @@ create_table_stmt:
...
+
+s3params:
+    s3param
+    {
+        $$ = $1
+    }
+|   s3params ',' s3param
+    {
+        $$ = append($1, $3...)
+        fmt.Println("wangjian test1 is", $$)
```
Comment (iamlinjunhong) [inline pkg/sql/parsers/dialect/mysql/mysql_sql.y:3938]: remove

File: pkg/sql/compile/scope.go (hunk 18 above)
Comment (nnsgmsone) [reply pkg/sql/compile/scope.go:54]: > I need this param to get the param like filepath

Why not write an opeator to handle this?

File: pkg/sql/compile/compile.go [hunk 22]
```diff
@@ -340,6 +341,18 @@ func (c *Compile) compilePlanScope(n *plan.Node, ns []*plan.Node) ([]*Scope, err
 			bat.InitZsOne(1)
 		}
 		ds.DataSource = &Source{Bat: bat}
+		if n.NodeType == plan.Node_EXTERNAL_SCAN {
```
Comment (nnsgmsone) [inline pkg/sql/compile/compile.go:344]: How do you implement this and how do you distribute parallel loads?

**Review by nnsgmsone — CHANGES_REQUESTED** [review]

File: pkg/sql/colexec/external/external.go [hunk 23]
```diff
@@ -0,0 +1,820 @@
...
+	var bat *batch.Batch
+	var err error
+	if param.plh == nil {
+		param.plh, err = GetSimdcsvReader(param)
+		if err != nil {
+			return nil, err
+		}
+	}
+	plh := param.plh
+	plh.simdCsvLineArray, err = plh.simdCsvReader.Read(param.batchSize)
```
Comment (nnsgmsone) [inline pkg/sql/colexec/external/external.go:793]: Where is the read implementation? I can't find the corresponding code implementation

Comment (jianwan0214) [reply pkg/sql/colexec/external/external.go]: it is now in my branch code, I will push it later

**Review by nnsgmsone — APPROVED** [review]

**Review by nnsgmsone — APPROVED** [review]

**Review by fengttt — APPROVED** [review]

**Review by iamlinjunhong — APPROVED** [review]

**Review by yingfeng — APPROVED** [review]

**Review by aunjgr — APPROVED** [review]

**Review by ouyuanning — APPROVED** [review]

**Review by XuPeng-SH — APPROVED** [review]
lgtm

**Review by daviszhen — APPROVED** [review]

**Pull request merged**